### DISCUSSION TITLE
does anyone help to write file editor?then can work with home assistant zigbee2mqtt

### DISCUSSION THREAD

**Discussion by haozee-harrison** [Koenkk/zigbee2mqtt#22333] [Devices]
we have some tuya zigbee devices.but we don't know how to write the xx.js file,then can work with home assistant out of box..does anyone help to write ?Thank you for your attention.

**Comment by schauveau** [comment]
I can try to help but I am not an expert.

First you need to clarify your request because I can understand it in multiple ways:

[1] You are a manufacturer with new Zigbee devices that are not recognized by  Z2M.
[2] or you own some Zigbee devices that are not supported by Z2M.
[3] or you own some Zigbee devices that have incomplete support in Z2M and you want to improve Z2M.
[4] or you own some Zigbee devices that already have good support in Z2M and you want to improve the integration in Home Assistant.

In all cases, you should probably give as much details as possible about the device(s). 

Also, posting some Zigbee2MQTT logs could help.

**Comment by haozee-harrison** [comment]
we have a tuya presence sensor and a water timer.when paired it with home assistant. the manuafacturer show "unsupported." model is TS061.

water timer link :[aliexpress.com/item/1005006859662031.html](https://www.aliexpress.com/item/1005006859662031.html)

we have a similar device with presence sensor. the model no is ZG-205Z/A.
The similar tuya water timer model no is QT06_2

Info 2024-04-30 13:36:23Device '0xa4c138f7ec1716a3' joined
Info 2024-04-30 13:36:23MQTT publish: topic 'zigbee2mqtt/bridge/event', payload '{"data":{"friendly_name":"0xa4c138f7ec1716a3","ieee_address":"0xa4c138f7ec1716a3"},"type":"device_joined"}'
Info 2024-04-30 13:36:23Starting interview of '0xa4c138f7ec1716a3'
Info 2024-04-30 13:36:24MQTT publish: topic 'zigbee2mqtt/bridge/event', payload '{"data":{"friendly_name":"0xa4c138f7ec1716a3","ieee_address":"0xa4c138f7ec1716a3","status":"started"},"type":"device_interview"}'
Info 2024-04-30 13:36:24MQTT publish: topic 'zigbee2mqtt/bridge/event', payload '{"data":{"friendly_name":"0xa4c138f7ec1716a3","ieee_address":"0xa4c138f7ec1716a3"},"type":"device_announce"}'
Info 2024-04-30 13:36:24Successfully interviewed '0xa4c138f7ec1716a3', device has successfully been paired
Warning 2024-04-30 13:36:24Device '0xa4c138f7ec1716a3' with Zigbee model 'TS0601' and manufacturer name '_TZE204_nbkshs6k' is NOT supported, please follow https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html
Info 2024-04-30 13:36:24MQTT publish: topic 'zigbee2mqtt/bridge/event', payload '{"data":{"definition":null,"friendly_name":"0xa4c138f7ec1716a3","ieee_address":"0xa4c138f7ec1716a3","status":"successful","supported":null},"type":"device_interview"}'

**Reply by Hoops79** [reply]
I have just received a Haozee presence sensor with the _tze204_nbkshs6k fingerprint. I am only able to get the following datapoints:

1 - presence (output appears to be inverted)
12 - illuminance

Have you managed to make any progress with this sensor?

Thanks

**Comment by haozee-harrison** [comment]
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {

    {
       fingerprint: tuya.fingerprint('TS0601', ['_TZE204_qasjif9e', '_TZE204_ztqnh5cg']),
        model: 'HZ-PS01',
        vendor: 'TuYa',
        description: 'human Presence sensor',
        fromZigbee: [tuya.fz.datapoints],
        toZigbee: [tuya.tz.datapoints],
        exposes: [
            e.illuminance_lux(), e.presence(),
            e.numeric('fading_time', ea.STATE_SET).withValue(30).withValue(60).withValue(120).
                .withDescription('Fading time').withUnit('s'),
        ],
        meta: {
            tuyaDatapoints: [
                [1, 'presence', tuya.valueConverter.trueFalse1], 
                [104, 'illuminance_lux', tuya.valueConverter.raw],
                [102, 'fading_time', tuya.valueConverter.raw],
    ],
},

module.exports = definition;

**Reply by schauveau** [reply]
Yes that could be a good starting point assuming that the devices are using Tuya datapoints but not all tuya device works that way. 

Look for the supported Tuya devices in  https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/tuya.ts and especially the ones with 'TS0601` in their fingerprint.

Most of them are using  tuyaDatapoints like in your converter but others require Endpoint configuration. 

For now, let's assume that your devices are using tuyaDatapoints.   
Also, some human presence devices can be quite complex so let's start with the water valve. 

Using the values from your logs, you can create a first external converter file `water_valve.js` in your `.../zigbee2mqtt/data/` directory where your `configuration.yaml` should be located.  

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
   fingerprint: tuya.fingerprint('TS0601', ['_TZE204_nbkshs6k']),
    model: 'PLEASE_GIVE_ME_A_UNIQUE_NAME',
    vendor: 'TuYa',
    description: 'Smart Water Valve',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    configure: tuya.configureMagicPacket,
    exposes: [
        e.numeric('dp999', ea.STATE).withDescription('Unknown datapoint'),
    ],
    meta: {
        tuyaDatapoints: [
            [999, 'dp999', tuya.valueConverter.raw],
        ],
    },
};

module.exports = definition;
```

Remarks:
  - Your example converter did not have `configure: tuya.configureMagicPacket,` but it is present in most official converters.  
  - The datapoint 999 is just an example. 
  - It is possible to check the syntax of the external converter using `node data/water_valve.js`. It is recommended to do it after each modification before restarting Z2M. 
  
You also need to declare the external converter 
  - in the Z2M web interface (in Settings / External Converters) 
  - or by adding the following lines to your `configuration.yaml`

```
external_converters:                                                                                                                                                                                                                                                                                                                                                                                     
  - water_valve.js
```

Now restart zigbee2mqtt. You may have to re-pair the device since the `configure` part may only be executed during pairing.

If everything works as expected, Z2M  should start receiving datapoints from the device and the DEBUG log should contain some `Received` messages for the device. You may also have a lot of warnings or errors about unknown datapoints.  

The datapoint numbers can be obtained in the `commandDataReport` messages from cluster `manuSpecificTuya`. Here is one for datapoint 115. We only care about the `"dp"` and `"datatype"` fields:  
 
```
 [2024-04-30 10:16:40] debug:    z2m: Received Zigbee message from 'energy_meter', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[255,255,254,255],"type":"Buffer"},"datatype":2,"dp":115}],"seq":33536}' from endpoint 1 with groupID 0
```

If you add entries in `exposes` and  `tuyaDatapoints` then your published state should start filling up with values. 

Look in  https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/tuya.ts for example of `exposes` and `tuyaDatapoints` entries.

The possible datatypes are 

```
    raw: 0, // [ bytes ]
    bool: 1, // [0/1]
    number: 2, // [ 4 byte value ]
    string: 3, // [ N byte string ]
    enum: 4, // [ 0-255 ]
    bitmap: 5, // [ 1,2,4 bytes ] as bits
```

The argument  `ea.STATE` in `exposes` is for read-only datapoints. 

If you believe that the datapoint is read-write then use `ea.STATE_SET` instead. This is probably the case for datapoint representing the on/off state. 
 
Have fun

**Comment by schauveau** [comment]
I believe that `tuya.valueConverter.raw` and `e.numeric` should be good for datatypes `bool (1)`, `number (2)`, `enum (4)`, and `bitmap (5)`. 

If you have a datatype `string (3)` then use `e.text` instead.

I am not sure what would be a good default for `raw (0)`.

**Comment by schauveau** [comment]
A small remark: If you want to shown some code or logs in a comment then use the **triple backticks** syntax described [here](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-cod). You can even specify the programming   language to get nice colors.  
 
    ```javascript  
    SOME JAVASCRIPT CODE HERE
    ```

**Comment by schauveau** [comment]
I incorrectly assumed that "Powered by Tuya" was an indication that this was a Tuya device but this is not necessarily the case since the GiEX water timer are also compatible with Tuya.
  
So the device could very well be a rebranded `QT06_1` or `QT06_2`. 

You could try to edit the file `.../zigbee2mqtt/node_modules/zigbee-herdsman-converters/devices/giex.js` and add `_TZE204_nbkshs6k` to the fingerprints of  `QT06_1` or `QT06_2`.

If it works then fill a issue in zigbee2mqtt to have it added officially.

**Comment by haozee-harrison** [comment]
i try to copy the manufacturer: GIEX,Model no: QT06_2,it still won't work..the JS file is as below..

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
   fingerprint: tuya.fingerprint('TS0601',  ['__TZE204_uab532m0', '_TZE204_4fblxpma']),
    model: 'HZ-WT-02',
    vendor: 'TuYa',
    description: 'Smart Water Valve',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    configure: tuya.configureMagicPacket,
    exposes: [
        exposes: [
            ...exportTemplates.giexWaterValve.exposes,
            e.numeric(legacy.giexWaterValve.irrigationTarget, ea.STATE_SET)
                .withValueMin(0)
                .withValueMax(SECONDS_IN_12_HOURS)
                .withUnit('seconds or litres')
                .withDescription('Irrigation target, duration in seconds or capacity in litres (depending on mode), ' +
                    'set to 0 to leave the valve on indefinitely, ' +
                    'for safety reasons the target will be forced to a minimum of 10 seconds in duration mode'),
            e.numeric(legacy.giexWaterValve.cycleIrrigationInterval, ea.STATE_SET)
                .withValueMin(0)
                .withValueMax(SECONDS_IN_12_HOURS)
                .withUnit('sec')
                .withDescription('Cycle irrigation interval'),
        ],
        whiteLabel: [
            tuya.whitelabel(HZ-WT-02',  'Water valve', ['_TZE204_uab532m0']),
        ],
    },
};

module.exports = definition;

**Comment by haozee-harrison** [comment]
we need to write correctly for the parts.

exposes: [
        e.numeric('dp999', ea.STATE).withDescription('Unknown datapoint'),
    ],
    meta: {
        tuyaDatapoints: [
            [999, 'dp999', tuya.valueConverter.raw],

**Reply by schauveau** [reply]
Remark: Please use the triple backticks syntax to make the code mode readable.

So I assume that you are trying to create an external converter inspired from `QT06_2` in `devices/giex.ts`

There are multiple issues in your versions

1.  This is not valid javascript. There are multiple syntax errors. Do not forget to run `node` on the  javascript file to get a diagnostic.
2. `legacy` was not imported. 
3. `SECONDS_IN_12_HOURS` and `exportTemplates.exposes` are not exported by `giec.js` so they cannot be importer.
They have to be copied here. 

Here is an external converter that should work better.  I could not test it of course but it should be syntactically correct:

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const legacy = require('zigbee-herdsman-converters/lib/legacy');

// giex.js does not export SECONDS_IN_12_HOURS so make out own local copy
const SECONDS_IN_12_HOURS = 12*3600;

const definition = {
    fingerprint: tuya.fingerprint('TS0601', ['__TZE204_uab532m0', '_TZE204_4fblxpma']),
    model: 'HZ-WT-02',
    vendor: 'TuYa',
    description: 'Smart Water Valve',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    configure: tuya.configureMagicPacket,
    exposes: [
        // .exportTemplates.exposes is not exported by `giex.ts
        // so copy its content here.
        e.battery(), // changed 'ep' to 'e'. Not sure why `giex.ts` was making a difference.
        e.binary(legacy.giexWaterValve.state, ea.STATE_SET, 'ON', 'OFF')
            .withDescription('State'),
        e.enum(legacy.giexWaterValve.mode, ea.STATE_SET, ['duration', 'capacity'])
            .withDescription('Irrigation mode'),
        e.numeric(legacy.giexWaterValve.cycleIrrigationNumTimes, ea.STATE_SET)
            .withValueMin(0)
            .withValueMax(100)
            .withDescription('Number of cycle irrigation times, set to 0 for single cycle'),
        e.numeric(legacy.giexWaterValve.irrigationStartTime, ea.STATE)
            .withDescription('Last irrigation start time'),
        e.numeric(legacy.giexWaterValve.irrigationEndTime, ea.STATE)
            .withDescription('Last irrigation end time'),
        e.numeric(legacy.giexWaterValve.lastIrrigationDuration, ea.STATE)
            .withDescription('Last irrigation duration'),
        e.numeric(legacy.giexWaterValve.waterConsumed, ea.STATE)
            .withUnit('L')
            .withDescription('Last irrigation water consumption'),
        //////////////////////////////
        e.numeric(legacy.giexWaterValve.irrigationTarget, ea.STATE_SET)
            .withValueMin(0)
            .withValueMax(SECONDS_IN_12_HOURS)
            .withUnit('seconds or litres')
            .withDescription('Irrigation target, duration in seconds or capacity in litres (depending on mode), ' +
                             'set to 0 to leave the valve on indefinitely, ' +
                             'for safety reasons the target will be forced to a minimum of 10 seconds in duration mode'),
        e.numeric(legacy.giexWaterValve.cycleIrrigationInterval, ea.STATE_SET)
            .withValueMin(0)
            .withValueMax(SECONDS_IN_12_HOURS)
            .withUnit('sec')
            .withDescription('Cycle irrigation interval'),
    ],
    // Somehow, an error occurs when using whiteLabel.
    // The fingerprint should be enough so I removed it. 
    //  whiteLabel: [
    //      tuya.whitelabel('HZ-WT-02', 'Water valve', ['_TZE204_uab532m0']),
    //  ],

};

module.exports = definition;
```

**Comment by schauveau** [comment]
I missed something important! The giex converters are not using Tuya datapoints so my previous converter will produce nothing.  

You need to replace  
```  
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    configure: tuya.configureMagicPacket,
```
by the values taken from `exportTemplates.giexWaterValve` :
```
    onEvent: tuya.onEventSetLocalTime,
    fromZigbee: [legacy.fromZigbee.giexWaterValve],
    toZigbee: [legacy.toZigbee.giexWaterValve],
```
